### ISSUE TITLE
`@JsonSubTypes` declared on multiple interfaces of a class results in order-dependent resolution and outcome

### ISSUE THREAD

**Issue by dweiss** [FasterXML/jackson-databind#3681]
This is arguably a corner case and (again, arguably) a bug in our code, but I think it'd be nice if jackson could at least return a better message here. TL;DR; version is: if you have an interface extending from more interfaces, each of which declares its own JsonSubTypes, the deserialization can be different and depends on how javac/runtime orders these interfaces in bytecode.

This is a distilled repro:
```
public class JsonTypeIdTest {
  @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
  @JsonSubTypes({
    @JsonSubTypes.Type(name = "c_impl", value = C_Impl.class)
  })
  private interface A {}

  @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
  @JsonSubTypes({
    @JsonSubTypes.Type(name = "b_impl", value = B_impl.class),
  })
  private interface B {}

  // Note the order of declarations on inherited interfaces - it makes
  // the difference in how types are resolved here.
  private interface C extends B, A {}

  private static class C_Impl implements C {}
  private static class B_impl implements B {}

  private static class Clazz {
    @JsonProperty public C c;
  }

  @Test
  public void checkSanity() throws IOException {
    var mapper = ObjectMapperProvider.createBuilder().build();
    mapper.readValue(
        """
        {
          "c": {
            "type": "c_impl"
          }
        }
        """,
        Clazz.class);
  }
}
```

The 'c' field on Clazz declares an instance of C. This is an interface that extends from A and B, each of which declare their implementing types. If you declare C to extend B, A, the above fails for me with:
```
com.fasterxml.jackson.databind.exc.InvalidTypeIdException: Could not resolve type id 'c_impl' as a subtype of `com.carrotsearch.lingo4g.internal.util.jackson.JsonTypeIdTest$C`: known type ids = [b_impl] (for POJO property 'c')
 at [Source: (String)"{
  "c": {
    "type": "c_impl"
  }
}
"; line: 3, column: 13] (through reference chain: com.carrotsearch.lingo4g.internal.util.jackson.JsonTypeIdTest$Clazz["c"])
```
so the type info is read from interface B only. But if you reorder the declaration to A, B, things will work just fine as "c_impl" type indeed implements interface C (and B).

I think the ideal way to solve the above would be to collect type mapping information from all superclasses/interfaces, then run a validation check whether type-class mappings are unique, if so - fine. I'm sure it'd break existing code here and there though...

**Comment by dweiss** [comment]
Thanks for confirming. I looked at the code briefly but it's (understandably), well, quite involving. :) I'll keep it in mind.

**Comment by cowtowncoder** [comment]
No problem. And thank you for including a reproduction; test case can make implementation little bit easier & ensure fix works.

**Comment by netomi** [comment]
Digging into that issue, it looks like the library can only handle a single annotation of a specific type at a time. Now when we collect all annotations of a class we look at the class itself and its super types. If an annotation is encountered that was already collected, it is silently ignored (see https://github.com/FasterXML/jackson-databind/blob/de62c677b12c02967c9620a11c1c5208f701cfef/src/main/java/com/fasterxml/jackson/databind/introspect/AnnotatedClassResolver.java#L301), leading to the situation that only the first encountered annotation is taken into account.

Thats why in the test case, "b_impl" is valid as it is the annotation for the first super type B. If we change the order to A, B, "c_impl" will be recognized instead.

The whole scenarios seems to be a bit constructed and I am not sure if this should be supported, but better error handling for scenarios that are not supported would be beneficial imho. There should be a way to indicate to the user that certain annotations are ignored by jackson. 

Are there already facilities that collect such information, or would make a strict mode make sense where jackson throws an IllegalArgumentException in such cases to fail-fast?

**Comment by cowtowncoder** [comment]
@netomi What we see is flattening of annotations so that any annotation introspection only needs to be concerned with the highest priority annotation for given processed class. This is something that is unlikely to ever be changed; requiring other parts of code base to try to figure out precedence in class hierarchies for annotations seems like huge additional load with limited value.

Although conflicting annotations might be something nice to catch, the problem is that flattening has absolutely no knowledge of semantics of various annotations, to know which are true conflicts, and if in some cases overrides are legit.

One area where things could possibly be improved would be ensuring proper override/precedence rules, if there are cases where existing rules could be improved. But from the original description I am not sure there is much as the suggestion is that in case of "equal" precedence, only one is eligible. But that would require code that flattens annotations to understand merging semantics -- and not just for annotations but with context (type hierarchy being resolved, base type etc).

So to be perfectly honest, I am not sure this case can be fully fixed.

**Comment by netomi** [comment]
Yeah, in order to support the reported use-case, jackson would need to merge collected JsonSubTypes annotations of all supertypes. This would require quite some change in the code as it is currently designed around a different assumption. Not really worth it imho, I was trying to figure out a way to warn the user that something unsupported is being configured, but as you say that would also be non-trivial and would have to take special cases into account.

**Comment by cowtowncoder** [comment]
@netomi Yes, I concur. On the other hand at least these are class-annotations (as opposed to field/method) so there'd be little bit less work -- but yes, some amount of information would need to be retained, and some amount of knowledge of merging aspect of specific annotation (meta-annotations?). In this case, definition could be along lines of reporting conflict if different Class value was found.
So maybe someone at some point could tackle this.

**Comment by cowtowncoder** [comment]
Merged in failing test to make it easier to verify improvements, if any. As per my comments I think error reporting is the thing to improve as usage itself might not be supportable (but maybe there's improvements there anyway; wrt merging of specific annotations).

**Comment by cowtowncoder** [comment]
Just confirming that this is unlikely to ever be fixed. Leaving open for now, however.

**Comment by dweiss** [comment]
Thank you, @cowtowncoder - I appreciate that the issue still gets your (and other people's) attention. I don't think there is harm in leaving this corner case as is. I believe this *could* be statically detected with tools like archunit, for example - it is more of a source code/ configuration problem than the framework's bug.
